### ISSUE TITLE
NavMeshSurface.BuildNavMesh() crash Unity with scenes loaded from asset bundles

### ISSUE THREAD

**Issue by KospY** [Unity-Technologies/NavMeshComponents#203]
Hi, I'm trying to build a navpath at runtime but unfortunately this plugin crash Unity as soon as I call NavMeshSurface.BuildNavMesh() with scenes loaded from asset bundles. 
Calling the same thing on non bundled scene work, but I got some errors about read / write, so I'm currently exploring this to see if it fix the problem. Anyway I think it's not normal that Unity crash so I prefer to open this issue while I'm investigating.
 
I'm using Unity 2019.4.20 LTS with the latest version of NavMeshComponents.

**Comment by KospY** [comment]
Update: I created a repro project. I can reproduce 100% the issue with our scene bundles.
Weirdly I created a bundle from the default Unity scene and it seem to work find with this one (but still throw error about read / write, so it confirm it's not related to this).

All our scene are crashing, I tried to pinpoint what causing this, deleting everything in the scene and calling BuildNavMesh work, but as soon as I let only one meshRenderer component it crash.

At this point I don't know what else I could do so I'm sending a bug report to Unity with the repro project and one of our scene bundle.

**Comment by turadr** [comment]
>  I'm sending a bug report to Unity with the repro project and one of our scene bundle.

Perfect! That's the right thing to do. Thank you.

**Comment by KospY** [comment]
Here is the reply from Unity QA:
`Thanks for getting in touch, we actually know about this issue and are tracking progress here: https://issuetracker.unity3d.com/product/unity/issues/guid/1298356`

We are not sure it's the same issue because baking work from editor, but once we export and import the scene as bundle it doesn't work anymore (it crash).
Anyway, does there is any workaround? We are doing randomized areas for our next update and we can't use any AI because of this bug. We are stuck right now and as issue is already open since 3 months we are a bit worried :/

**Comment by turadr** [comment]
The two issues are very similar because they crash in the same section of code, more or less. We haven't looked yet at the issue you've reported, but case 1298456 is currently under investigation. Preliminary information suggests that it is caused by meshes that are placed very far away from the origin, in conjunction with a very small value for the "voxel size" parameter.  The code for building the NavMesh cannot handle well the vertices that are very far away from the origin due to limitations in the floating point precision.

As a workaround you could maybe try to limit the distance up to where you build the NavMesh, either by [not using far-away meshes](https://github.com/Unity-Technologies/NavMeshComponents/blob/master/Documentation/NavMeshModifier.md#parameters) as a source, or by choosing a smaller [local bounding volume](https://github.com/Unity-Technologies/NavMeshComponents/blob/master/Documentation/NavMeshSurface.md#parameters).

**Comment by KospY** [comment]
That's weird because we are really close to the origin, on the example scene I attached to the bug report nothing is further than 150 units, it's really a simple scene. Voxel size is set to auto and show 0.13.

**Comment by KospY** [comment]
I can confirm that it has nothing to do with meshes or terrain patches that are placed very far away from the origin.
I spent the whole day working on it and finally found out that only a few meshes can cause a crash. This meshes have nothing special, are placed near the origin and scaled normally.

What is causing the crash is in fact batching static. Once checked, if I build the navmesh in the editor, it will warn about the combined mesh not being able to be read and generate a wrong navmesh. If I bundle the scene from our project, load the bundle in another editor and then build the navmesh it will crash. Bundling the same scene on the same project work for some reason however, but the navmesh is still wrong. We are using addressable for our project so maybe the bundles are packed differently?

I also found that if I disable batching static, and let the mesh with read/write unchecked, Unity still warns about the mesh but the navmesh will be built correctly and will also not crash if loaded from a bundle.

I sent another mail to Unity support about that. Meanwhile, our level artist doing the dungeon rooms will need to be vigilant because if any mesh have batching static enabled, it will crash the game :(

**Comment by tsantarra** [comment]
Hello, @KospY! 

I believe you are correct. My investigation of your project had led me to an existing issue ([link](https://github.com/Unity-Technologies/NavMeshComponents/issues/58)), which similarly identifies static batching as the cause. Static batching is not supported by navmesh at this time. 

We initially suspected the loaded meshes were far from the origin due to the vertex data being operated on at the time of the crash. The vertices were seemingly millions of units away. In truth, uninitialized memory had been fed into the navmesh build process.

**Comment by KospY** [comment]
After more tests it's not only because of batching static. We have some rooms that constantly crash even with batching static unchecked.
I'm spending hours disabling / enabling meshes, crash, restart unity, rebundle, etc... But I can't find the logic of this crash.
It seem to happen with a combination of some meshes. This meshes have nothing particular, they are not oversized or far from origin. Something is broken with addressable bundles and navpath.

As we can't find a workaround, we are looking to change to a third party solution for navpath, it mean reworking our AI pathfinding but I think it would be quicker and safer than waiting Unity to fix the issue :(

**Comment by tsantarra** [comment]
I noticed in the repro project that some of the meshes report positive vertex counts but have no associated data (null pointer). My intuition is that either something in the bundling process is causing this or some of the source meshes incorporated into the bundle have been corrupted somehow.

**Comment by KospY** [comment]
Yeah it seem that meshes get bundled differently from our project for some reason. When I build the same scene from an empty project using asset bundles it worked fine. Maybe it's addressable doing something different.

Anyway we finally found a workaround! 
We pre-baked each room in the editor, and then added NavMeshLink between each rooms at runtime. This worked fine as we don't have to call BuildNavMesh()! :)
